### PR TITLE
If Out of Range, recalculate the price range using the current price ratio

### PR DESCRIPTION
**Pull request by joaobrunoah**
# Description

This PR has 2 changes:
1. When the pool is out of range and we recompute the virtual balances, we use the current price ratio to avoid virtual balances jumping and math underflows (the function `computeVirtualBalancesUpdatingPriceRange` was conceived to use with a constant price ratio).
2. An E2E test with high fees and high volume forces the pool to collect a lot of LP fees, which raises the price ratio by almost 4x. Then, creates swaps that recomputes the virtual balances. These tests would fail without the changes from [1].

## Type of change

- [x] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Optimization: [ ] gas / [ ] bytecode
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests have 100% code coverage
- [x] The base branch is either `main`, or there's a description of how to merge

### CODE REVIEW

**Review by EndymionJkb — APPROVED** [review]
LGTM

File: test/foundry/E2eSwapReClammSwapFees.t.sol [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                0,
+                MAX_UINT256,
+                false,
+                bytes("")
+            );
+            router.swapSingleTokenExactOut(
+                pool,
+                tokenB,
+                tokenA,
+                (poolInitAmountTokenA) / 300,
```
Comment (jubeira) [inline test/foundry/E2eSwapReClammSwapFees.t.sol:104]: Why are we dividing by 300 here instead of 3?

File: test/foundry/E2eSwapReClammSwapFees.t.sol [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+import { PoolConfigLib } from "@balancer-labs/v3-vault/contracts/lib/PoolConfigLib.sol";
+import { E2eSwapTest } from "@balancer-labs/v3-vault/test/foundry/E2eSwap.t.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { ReClammPool } from "../../contracts/ReClammPool.sol";
+import { ReClammMath, a, b } from "../../contracts/lib/ReClammMath.sol";
+import { ReClammPoolMock } from "../../contracts/test/ReClammPoolMock.sol";
+import { E2eSwapFuzzPoolParamsHelper } from "./utils/E2eSwapFuzzPoolParamsHelper.sol";
+
+contract E2eSwapReClammSwapFeesTest is E2eSwapFuzzPoolParamsHelper, E2eSwapTest {
```
Comment (jubeira) [inline test/foundry/E2eSwapReClammSwapFees.t.sol:25]: Fair enough.

I think we should also test "the jump" more explicitly. I.e. after the setup where the fees have accumulated, check:
- that the liquidity has deconcentrated (do we need to assume that [here](https://github.com/balancer/reclamm/pull/113/files#diff-7385a1e63b393af5ac194e414d27e60205d089f39c05a937cd820bb9edc767b8R117)? Doesn't it always happen?)
- check what happens to the spot price after a small swap out of range
- check what happens to virtual balances  / invariant after a small price out of range

Not sure if anything else. I'd just like to make sure that there's nothing weird around this edge case.

**Review by jubeira — COMMENTED** [review]
Changes make sense, the diff is pretty clean. Just a few questions for now; other than that looks good.

Can we double down on the tests a bit? I'd like to make sure that everything looks good after we update things when the pool gets out of range, and that the price doesn't jump. The current e2e test kinda does that, but not explicitly. In other words, how do we know that the underlying e2e swaps check what happens when the pool is out of range?

File: test/foundry/E2eSwapReClammSwapFees.t.sol (hunk 1 above)
Comment (joaobrunoah) [reply test/foundry/E2eSwapReClammSwapFees.t.sol]: I added an explanation in the code. The idea is to make sure the amount in of token B is equivalent to the amount in of token A, to don't take the pool very far from the balances we want to test.

Comment (jubeira) [reply test/foundry/E2eSwapReClammSwapFees.t.sol]: can't we just use the amount calculated in the first step and put it back in the pool?

File: contracts/lib/ReClammMath.sol [hunk 3]
```diff
@@ -501,8 +497,9 @@ library ReClammMath {
         uint32 currentTimestamp,
         uint32 lastTimestamp
     ) internal pure returns (uint256 newVirtualBalanceA, uint256 newVirtualBalanceB) {
-        // Round up price ratio, to round virtual balances down.
-        uint256 priceRatio = currentFourthRootPriceRatio.mulUp(currentFourthRootPriceRatio);
+        uint256 sqrtPriceRatio = Math.sqrt(
+            computePriceRatio(balancesScaled18, virtualBalanceA, virtualBalanceB) * 1e18
```
Comment (elshan-eth) [inline contracts/lib/ReClammMath.sol:501]: 1e18 -> FixedPoint.One

File: contracts/lib/ReClammMath.sol [hunk 4]
```diff
@@ -616,6 +613,64 @@ library ReClammMath {
...
+     * These calculations are derived from the invariant equation: invariant = (Ra + Va)(Rb + Vb),
+     * where Ra and Rb are the real balances of tokens A and B respectively.
+     *
+     * @param balancesScaled18 Current pool balances, sorted in token registration order
+     * @param virtualBalanceA Virtual balance of token A
+     * @param virtualBalanceB Virtual balance of token B
+     * @return minPrice The minimum price of token A in terms of token B
+     * @return maxPrice The maximum price of token A in terms of token B
+     */
+    function computePriceRange(
```
Comment (elshan-eth) [inline contracts/lib/ReClammMath.sol:649]: It might be worth adding TypeScript tests for these methods in reClammMath.test.ts

File: test/gas/.hardhat-snapshots/[ReClamm - OOR - Standard - BatchRouter] swap exact in with one token and fees - cold slots [hunk 5]
```diff
@@ -1 +1,5 @@
-217.7k
\ No newline at end of file
+<<<<<<< HEAD
```
Comment (elshan-eth) [inline test/gas/.hardhat-snapshots/[ReClamm - OOR - Standard - BatchRouter] swap exact in with one token and fees - cold slots:1]: Looks like the snapshots need to be updated

**Review by elshan-eth — APPROVED** [review]
Awesome! Really great work! 
Just small comments and I agree with @jubeira comments about the additional tests.

File: test/foundry/E2eSwapReClammSwapFees.t.sol (hunk 2 above)
Comment (joaobrunoah) [reply test/foundry/E2eSwapReClammSwapFees.t.sol:25]: Done! There's a hardhat test now that takes deconcentrates the liquidity through swaps, then move pool out of range, then updates price ratio, then does a final swap. Prices are checked on each step, to make sure they make sense.

We can't really compare invariants, it'll decrease since the price range is moving, but at least we know the prices are not jumping.

File: test/foundry/ReClammMath.t.sol [hunk 6]
```diff
@@ -553,6 +553,29 @@ contract ReClammMathTest is BaseReClammTest {
         assertEq(mathPow4, fpPow4, "Pow4 value mismatch");
     }
 
+    function testComputePriceRange__Fuzz(uint256 minPrice, uint256 maxPrice, uint256 targetPrice) public pure {
```
Comment (jubeira) [inline test/foundry/ReClammMath.t.sol:556]: nice

File: test/reClammPool.test.ts [hunk 7]
```diff
@@ -315,4 +328,396 @@ describe('ReClammPool', function () {
...
+          false,
+          '0x'
+        );
+    }
+
+    // 0% swap fee, making sure no fees will be accrued by the pool in the next swaps.
+    await vault.connect(bob).manualUnsafeSetStaticSwapFeePercentage(pool, fp(0));
+
+    const initialFourthRootPriceRatio = await pool.computeCurrentFourthRootPriceRatio();
+    // Make sure the price ratio raised by 16 (2^4), at least.
```
Comment (jubeira) [inline test/reClammPool.test.ts:371]: You mean price ratio did 2x, and fourth root price ratio did 16x?

File: test/reClammPool.test.ts [hunk 8]
```diff
@@ -315,4 +328,396 @@ describe('ReClammPool', function () {
...
+          fpMulDown(INITIAL_BALANCE_A, fp(0.9 * 0.8)),
+          MAX_UINT256,
+          MAX_UINT256,
+          false,
+          '0x'
+        );
+    }
+
+    // 0% swap fee, making sure no fees will be accrued by the pool in the next swaps.
+    await vault.connect(bob).manualUnsafeSetStaticSwapFeePercentage(pool, fp(0));
```
Comment (jubeira) [inline test/reClammPool.test.ts:554]: Perhaps you can move this block inside a `sharedBeforeEach`?

Comment (jubeira) [reply test/reClammPool.test.ts:554]: (or even include a few more lines)

**Review by jubeira — COMMENTED** [review]
Very nice. This looks pretty good!

I'd like to take one more deep look at the hardhat test before proceeding; there's a lot going on in there as expected. Great job!

File: test/reClammPool.test.ts (hunk 7 above)
Comment (joaobrunoah) [reply test/reClammPool.test.ts:371]: If fourth root price ratio raises by a factor of 2, the price ratio raises by a factor of 16.

File: test/reClammPool.test.ts [hunk 9]
```diff
@@ -46,11 +52,13 @@ describe('ReClammPool', function () {
...
-  const PRICE_SHIFT_DAILY_RATE = fp(1); // 100%. Price interval can double or reduce by half each day
-  // 20%. If pool centeredness is less than margin, price interval will track the market price.
-  const CENTEREDNESS_MARGIN = fp(0.2);
+  // 100%. Price interval can double or reduce by half each day.
+  const PRICE_SHIFT_DAILY_RATE = fp(1);
+  // 50%. If pool centeredness is less than margin, price interval will track the market price.
+  const CENTEREDNESS_MARGIN = fp(0.5);
 
   const virtualBalancesError = 0.000000000000001;
+  const pricesError = 0.06; // 6% error tolerance.
```
Comment (jubeira) [inline test/reClammPool.test.ts:61]: This is a bit high; doesn't look like a rounding error.
Given that we use the same error across the test, I can't tell where exactly you need the number to be this big. Could you explain the discrepancy between expected and calculated prices?

Comment (jubeira) [reply test/reClammPool.test.ts:61]: E.g. here: https://github.com/balancer/reclamm/pull/113/files#diff-d8c53d21aa60f6d47e8cf5c0c9bad3d4c391299b1b3da9041e24a6c35063a598R395 we're using the prices before a swap as expected prices after a big swap, and swaps do have price impact. Isn't the 6% error masking the price impact in this case?

File: test/reClammPool.test.ts (hunk 7 above)
Comment (jubeira) [reply test/reClammPool.test.ts:371]: :sweat_smile: lol that's what I meant, thanks.

File: test/reClammPool.test.ts (hunk 9 above)
Comment (joaobrunoah) [reply test/reClammPool.test.ts:61]: I'll expose the error in the function, and set this high tolerance only when it's needed. It's usually needed after the 6 hours warp, because my price calculation in the test is just an approximation of where the price really should be.

Comment (joaobrunoah) [reply test/reClammPool.test.ts]: Tests are fixed, now the precision is clearer.

File: test/reClammPool.test.ts (hunk 8 above)
Comment (joaobrunoah) [reply test/reClammPool.test.ts]: Done!

File: contracts/lib/ReClammMath.sol (hunk 4 above)
Comment (joaobrunoah) [reply contracts/lib/ReClammMath.sol:649]: Done!

File: test/reClammPool.test.ts [hunk 10]
```diff
@@ -315,4 +337,818 @@ describe('ReClammPool', function () {
...
+    expectEqualWithError(fpDivDown(maxPrice, minPrice), priceRatio, expectedPriceRatioError);
+    if (compareMinAndMaxPrices) {
+      expectEqualWithError(minPrice, expectedMinPrice, expectedPricesError);
+      expectEqualWithError(maxPrice, expectedMaxPrice, expectedPricesError);
+    }
+
+    return { minPrice, maxPrice };
+  }
+
+  async function checkSpotPriceAfterSwap(virtualBalancesBeforeSwap: bigint[]) {
```
Comment (jubeira) [inline test/reClammPool.test.ts:1136]: Can we add a small comment (either here or in the function call) that explains why we're warping the time and making this check?

File: test/reClammPool.test.ts [hunk 11]
```diff
@@ -315,4 +337,818 @@ describe('ReClammPool', function () {
...
+      const deadline = MAX_UINT256;
+      const wethIsEth = false;
+
+      const virtualBalancesBeforeSwap = await pool.computeCurrentVirtualBalances();
+
+      await router
+        .connect(bob)
+        .swapSingleTokenExactOut(pool, tokenB, tokenA, exactAmountOut, maxAmountIn, deadline, wethIsEth, '0x');
+
+      await checkSpotPriceAfterSwap(virtualBalancesBeforeSwap);
```
Comment (jubeira) [inline test/reClammPool.test.ts:372]: Hm virtual balances should stay the same before and after this swap right? Why do we need to store them before calling swap?

Perhaps one good check would be to verify that the virtual balances actually stay the same before and after the swap (since the timestamp is not moving, and also the swap happens in range).

**Review by jubeira — COMMENTED** [review]
Just a few more questions and minor observations, otherwise LGTM! Great job @joaobrunoah

File: test/reClammPool.test.ts [hunk 12]
```diff
@@ -315,4 +337,818 @@ describe('ReClammPool', function () {
...
+
+      // Concentrating liquidity
+      // Since the price move introduces some rounding, store the price ratio before the setPriceRatioState call.
+      // Notice that "checkPoolPrices" already checked that initialFourthRootPriceRatio matches the current price ratio,
+      // so the values are close.
+      const startFourthRootPriceRatio = await pool.computeCurrentFourthRootPriceRatio();
+      const updateStartTimestamp = (await currentTimestamp()) + 1n;
+      const updateEndTimestamp = updateStartTimestamp + 1n * BigInt(DAY) + 1n;
+      const endFourthRootPriceRatio = fpDivDown(initialFourthRootPriceRatio, fp(1.1));
+      await pool.connect(bob).setPriceRatioState(endFourthRootPriceRatio, updateStartTimestamp, updateEndTimestamp);
```
[the remark this answers is not in the document]
Comment (joaobrunoah) [reply test/reClammPool.test.ts:412]: The endFourthRootPriceRatio is calculated by divDown, so I'm decreasing the fourthRootPriceRatio, oncentrating the liquidity.

File: test/reClammPool.test.ts [hunk 13]
```diff
@@ -315,4 +337,818 @@ describe('ReClammPool', function () {
...
+
+      // Deconcentrating liquidity
+      // Since the price move introduces some rounding, store the price ratio before the setPriceRatioState call.
+      // Notice that "checkPoolPrices" already checked that initialFourthRootPriceRatio matches the current price ratio,
+      // so the values are close.
+      const startFourthRootPriceRatio = await pool.computeCurrentFourthRootPriceRatio();
+      const updateStartTimestamp = (await currentTimestamp()) + 1n;
+      const updateEndTimestamp = updateStartTimestamp + 1n * BigInt(DAY) + 1n;
+      const endFourthRootPriceRatio = fpMulDown(initialFourthRootPriceRatio, fp(1.1));
+      await pool.connect(bob).setPriceRatioState(endFourthRootPriceRatio, updateStartTimestamp, updateEndTimestamp);
```
[the remark this answers is not in the document]
Comment (joaobrunoah) [reply test/reClammPool.test.ts:768]: I just chose a value, but I can concentrate in the maximum amount that I can, which given the guardrails is 1.68 (which leads to a price ratio of 8).

File: test/reClammPool.test.ts (hunk 10 above)
Comment (joaobrunoah) [reply test/reClammPool.test.ts:1140]: Done!

File: test/reClammPool.test.ts [hunk 14]
```diff
@@ -315,4 +337,818 @@ describe('ReClammPool', function () {
...
+        pool,
+        initialFourthRootPriceRatio,
+        minPriceBeforeBigSwap,
+        maxPriceBeforeBigSwap,
+        priceRatioError,
+        pricesSmallError,
+        true
+      );
+
+      await advanceTime(6 * HOUR);
```
[the remark this answers is not in the document]
Comment (joaobrunoah) [reply test/reClammPool.test.ts:360]: Yes, but then I'm not checking the speed of change. I mean, if I jump 1 day and check if it doubled, I'm not sure if it was linear, or if the curve of the price ratio and the update of virtual balances is behaving as we predict. But jumping 6 hours and making these approximate assertions makes me more confident that the pool is behaving as we predict.

File: test/reClammPool.test.ts [hunk 15]
```diff
@@ -315,4 +337,818 @@ describe('ReClammPool', function () {
...
+      const exactAmountOut = fpMulDown(poolBalancesBeforeSwapRaw[tokenAIdx], fp(0.99));
+      const maxAmountIn = MAX_UINT256;
+      const deadline = MAX_UINT256;
+      const wethIsEth = false;
+
+      const virtualBalancesBeforeSwap = await pool.computeCurrentVirtualBalances();
+
+      await router
+        .connect(bob)
+        .swapSingleTokenExactOut(pool, tokenB, tokenA, exactAmountOut, maxAmountIn, deadline, wethIsEth, '0x');
```
[the remark this answers is not in the document]
Comment (joaobrunoah) [reply test/reClammPool.test.ts]: With the new function, it's clearer. Still, I wrote a few more information in the swap comment.

File: test/reClammPool.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (jubeira) [reply test/reClammPool.test.ts:697]: Please disregard all these comments; my bad :)

File: test/reClammPool.test.ts [hunk 16]
```diff
@@ -1133,16 +1029,65 @@ describe('ReClammPool', function () {
...
+
+    // Makes sure swap and virtual balances transactions are batched in the same block, so the timestamp does not
+    // change.
+    await ethers.provider.send('evm_setAutomine', [false]);
+
+    await router
+      .connect(bob)
+      .swapSingleTokenExactOut(pool, tokenIn, tokenOut, exactAmountOut, maxAmountIn, deadline, wethIsEth, '0x');
+
+    const virtualBalancesAfterSwap = await pool.computeCurrentVirtualBalances();
```
[the remark this answers is not in the document]
Comment (jubeira) [reply test/reClammPool.test.ts:1055]: got it, makes sense then!
